doc: Add Azure DevOps OAuth provider docs (#48822)
sourcegraph/sourcegraph: doc/admin/config/authorization_and_authentication.md

Follow the steps to [sync Bitbucket Server / Bitbucket Data Center permissions](../external_service/bitbucket_server.md#repository-permissions). Then, do one of the following:

1. Create the user accounts in Sourcegraph with matching usernames. (Access using `builtin` auth.)
2. [Configure SAML authentication](../auth/saml/index.md). If you are using Bitbucket Server / Bitbucket Data Center, the `login` attribute is _not_ optional. You need to pass the Bitbucket Server username as the `login` attribute.
3. [Configure an HTTP authentication proxy](../auth/index.md#http-authentication-proxies), passing the Bitbucket Server username value as the `usernameHeader`.

### Azure DevOps Services

> We support authentication through OAuth for Azure DevOps Services (https://dev.azure.com) and it is also a prerequisite for permissions syncing.

Reviewer: ```suggestion
We support authentication through OAuth for [Azure DevOps Services](https://dev.azure.com) and it is also a prerequisite for [permissions syncing](../permissions/index.md).
```

It's Markdown after all :)

Author: I wanted to make the URL explicit here to remove any confusion while reading instead of the user having to hover. Does it make more sense given the context?

Reviewer: Then maybe `[Azure DevOps Services (dev.azure.com)](https://dev.azure.com)`?

Author: Hmm, agreed. That would make sense.

#### Register a new OAuth app

> Visit the URL [here](https://app.vsaex.visualstudio.com/app/register) and follow the instructions below:

Reviewer: See: https://handbook.sourcegraph.com/company-info-and-process/communication/content_guidelines/actionable_language/#links

We shouldn't use `click [here]`, but instead to something like

```suggestion
1. Start the [registration process for a new Azure DevOps application](https://app.vsaex.visualstudio.com/app/register)
```

1. In the `Application website` field set the URL of your Sourcegraph instance, for example if the instance is https://sourcegraph.com, then use `https://sourcegraph.com` as the value of this field
2. Similarly, set the `Authorization callback URL` field to `https://sourcegraph.com/.auth/azuredevops/callback` if your Sourcegraph instance URL is https://sourcegraph.com
3. Add the following scopes:
   - `User profile (read)`
   - `Identity (read)`
   - `Code (read)`
   - `Project and team (read)`

#### Configuring Sourcegraph auth.providers

Add the following to the `auth.providers` key in the site config:

> ```

Reviewer: ```suggestion
```json
```

{
  "auth.providers": [
    // Other auth providers may also be here.
    {
      "type": "azureDevOps",
      "displayName": "Azure DevOps",

>       "clientID": $APP_ID_FROM_THE_PREVIOUS_STEP,

Reviewer: @varsanojidan @indradhanush I ran into this while setting it up locally: it's super confusing that it's called `clientID` and not `appID`. I went back twice to check whether I don't have a client ID, even though it says here `$APP_ID`.

Reviewer: I now realise that the steps below the configuration call this out, but at this point I was already lost. I suggest we start with "Here's the information you need" in this section and move the list from line 164-167 up to 144 or something like that.

      "clientSecret": $CLIENT_SECRET,
      "apiScope": "vso.code,vso.identity,vso.project"
    },
  ]
}
```

It is **important** to ensure that:

1. The value of `App ID` from your OAuth application is set as the value of the `clientID` field
2. The value of `Client Secret` (and not the `App secret`) from your OAuth application is set as the value of the `clientSecret` field
3. The value of `apiScope` string is a comma separated string and reflects the scopes from your OAuth application accurately
4. The `type` field has no typos and is **exactly** the same as the example above

Optionally, you may want to restrict the sign up to only users who belong to a specific list of organizations. To do this add the following to the `auth.providers` configuration above:

```
      "allowOrgs": ["your-org-1", "your-org-2"],
```

Finally, if you want to prevent new users from signing up to your Sourcegraph instance, set the following (default to `true`) in the `auth.providers` configuration above:

```
      "allowSignup": false
```

The final and complete `auth.providers` configuration may look like this:

> ```

Reviewer: ```suggestion
```json
```
